File: src/leetcodepython/string/group_anagrams_49.py
# encoding='utf-8'

'''
/**
 * This is the solution of No. 49 problem in the LeetCode,
 * the website of the problem is as follow:
 * https://leetcode-cn.com/problems/group-anagrams
 *
 * The description of problem is as follow:
 * ==========================================================================================================
 * 给定一个字符串数组，将字母异位词组合在一起。字母异位词指字母相同，但排列不同的字符串。
 *
 * 示例:
 *
 * 输入: ["eat", "tea", "tan", "ate", "nat", "bat"]
 * 输出:
 * [
 *   ["ate","eat","tea"],
 *   ["nat","tan"],
 *   ["bat"]
 * ]
 * 说明：
 *
 * 所有输入均为小写字母。
 * 不考虑答案输出的顺序。
 *
 * 来源：力扣（LeetCode）
 * 链接：https://leetcode-cn.com/problems/group-anagrams
 * 著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。
 * ==========================================================================================================
 *
 * @author  zhangyu (zhangyuyu417@gmail.com)
 */
'''
from typing import List


class Solution:
    def group_anagrams1(self, strs: List[str]) -> List[str]:
        '''
            字符串
        Args:
            strs: 字符串数组
        Returns:
            链表
        '''
        str_dict = {}
        for ele in strs:
            sorted_str = self.get_sort_str(ele)
            if sorted_str not in str_dict:
                str_dict[sorted_str] = []
            str_dict[sorted_str].append(ele)
        return list(str_dict.values())

    def get_sort_str(self, s: str) -> str:
        '''
            对字符串进行排序
        Args:
            s: 字符串
        Returns:
            排序后字符串
        '''
        if not s or len(s) < 1:
            return None
        s = sorted(list(s))
        return ''.join(s)

    def group_anagrams2(self, strs: List[str]) -> List[str]:
        '''
            字符串
        Args:
            strs: 字符串数组
        Returns:
            链表
        '''
        words_dict = {}
        for str in strs:
            word_key = self.word_to_num(str)
            if word_key not in words_dict:
                words_dict[word_key] = []
            words_dict[word_key].append(str)
        return list(words_dict.values())

    def word_to_num(self, s: str) -> int:
        '''
            将字符串转成素数
        Args:
            s:将字符串转成数字
        Returns:
            质数
        '''
        # 26个字母，定义26个素数
        prime = [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97, 101, 103]
        key = 1
        for ch in s:
            key *= prime[ord(ch) - ord('a')]
        return key


if __name__ == '__main__':
    words = ["eat", "tea", "tan", "ate", "nat", "bat"]
    solution = Solution()
    result = solution.group_anagrams1(words)
    print(result)
    assert result == [['eat', 'tea', 'ate'], ['tan', 'nat'], ['bat']]

File: src/leetcodejava/string/GroupAnagrams49.java
package leetcodejava.string;

import org.junit.Assert;
import org.junit.Test;

import java.util.*;

/**
 * This is the solution of No. 49 problem in the LeetCode,
 * the website of the problem is as follow:
 * https://leetcode-cn.com/problems/group-anagrams
 * <p>
 * The description of problem is as follow:
 * ==========================================================================================================
 * 给定一个字符串数组，将字母异位词组合在一起。字母异位词指字母相同，但排列不同的字符串。
 * <p>
 * 示例:
 * <p>
 * 输入: ["eat", "tea", "tan", "ate", "nat", "bat"]
 * 输出:
 * [
 * ["ate","eat","tea"],
 * ["nat","tan"],
 * ["bat"]
 * ]
 * 说明：
 * <p>
 * 所有输入均为小写字母。
 * 不考虑答案输出的顺序。
 * <p>
 * 来源：力扣（LeetCode）
 * 链接：https://leetcode-cn.com/problems/group-anagrams
 * 著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。
 * ==========================================================================================================
 *
 * @author zhangyu (zhangyuyu417@gmail.com)
 */
public class GroupAnagrams49 {
    @Test
    public void groupAnagramsTest() {
        String[] strs = {"aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa", "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"};
        List<List<String>> result = groupAnagrams4(strs);
        System.out.println(result);
        Assert.assertEquals(result.get(0).get(0), "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa");
    }


    /**
     * 生成镶嵌链表
     *
     * @param strs 字符串数组
     * @return 链表
     */
    private List<List<String>> groupAnagrams2(String[] strs) {
        int[] prime = {2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97, 101, 103};//最多10609个z
        List<List<String>> res = new ArrayList<>();
        HashMap<Integer, Integer> map = new HashMap<>();
        for (String s : strs) {
            int key = 1;
            for (char c : s.toCharArray()) {
                key *= prime[c - 'a'];
            }
            List<String> t;
            if (map.containsKey(key)) {
                t = res.get(map.get(key));
            } else {
                t = new ArrayList<>();
                res.add(t);
                map.put(key, res.size() - 1);
            }
            t.add(s);
        }
        return res;
    }

    /**
     * 生成镶嵌链表
     *
     * @param strs 字符串数组
     * @return 链表
     */
    private List<List<String>> groupAnagrams4(String[] strs) {
        if (strs == null || strs.length < 1) {
            return new ArrayList<>();
        }
        int[] prime = {2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97, 101, 103};//最多10609个z
        HashMap<Double, List<String>> map = new HashMap<>();
        for (String s : strs) {
            double key = 1;
            for (char c : s.toCharArray()) {
                key *= prime[c - 'a'];
            }
            if (map.containsKey(key)) {
                map.get(key).add(s);
            } else {
                List<String> t = new ArrayList<>();
                t.add(s);
                map.put(key, t);
            }
        }
        return new ArrayList<>(map.values());
    }

    /**
     * 生成镶嵌链表
     *
     * @param strs 字符串数组
     * @return 链表
     */
    private List<List<String>> groupAnagrams3(String[] strs) {
        if (strs == null || strs.length < 1) {
            return new ArrayList<>();
        }
        Map<String, List<String>> map = new HashMap<>();
        for (String str : strs) {
            char[] array = str.toCharArray();
            Arrays.sort(array);
            String key = new String(array);
            if (map.containsKey(key)) {
                map.get(key).add(str);
            } else {
                map.put(key, new ArrayList<>(Arrays.asList(str)));
            }
        }
        return new ArrayList<>(map.values());
    }

    /**
     * 生成镶嵌链表
     *
     * @param strs 字符串数组
     * @return 链表
     */
    private List<List<String>> groupAnagrams1(String[] strs) {
        Map<String, List<String>> map = new HashMap<>();
        for (String str : strs) {
            String newStr = getSortedStr(str);
            if (!map.containsKey(newStr)) {
                map.put(newStr, new ArrayList<>());
            }
            map.get(newStr).add(str);
        }
        return new ArrayList<>(map.values());
    }

    /**
     * 字符串进行排序
     *
     * @param s 字符串
     * @return 排序好的字符串
     */
    public String getSortedStr(String s) {
        char chs[] = s.toCharArray();
        Arrays.sort(chs);
        return new String(chs);
    }
}
